Move masonry logic from RenderGrid to GridMasonryLayout.

sammygill opened the pull request:
#### ee553c43b364439656fe9eb00616b8167c8d103a
<pre>
Move masonry logic from RenderGrid to GridMasonryLayout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248621">https://bugs.webkit.org/show_bug.cgi?id=248621</a>
rdar://102871735

Reviewed by Tim Nguyen.

This is a somewhat small refactor that involes moving some masonry
specific logic from RenderGrid to GridMasonryLayout. The intial goal was
to remove only collectMasonryItems so that GridMasonryLayout could
hold onto the items directly, but it became pretty clear that some of
the other logic in the surrounding area could also be moved. This should
hopefully have 2 benefits:

1.) Make RenderGrid much more simple and easier to read when we need to
do a masonry layout (since much of the responsibility should be moved)
2.) Allow GridMasonryLayout to much more easily query the masonry item
data structures without having to pass references around everywhere.
This is especially useful if we ever need to check if an item is a
first track item.

There should be no functionality change with this patch.

* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::performMasonryPlacement):
(WebCore::GridMasonryLayout::collectMasonryItems):
(WebCore::GridMasonryLayout::allocateCapacityForMasonryVectors):
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
(WebCore::GridMasonryLayout::nextMasonryPositionForItem):
(WebCore::GridMasonryLayout::gridAxisDirection const):
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::allocateSpaceForMasonryVectors): Deleted.
(WebCore::itemGridAreaStartsAtFirstLine): Deleted.
(WebCore::itemGridAreaIsWithinImplicitGrid): Deleted.
(WebCore::RenderGrid::collectMasonryItems const): Deleted.
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridPositionsResolver.h:
(WebCore::GridPositionsResolver::gridAxisDirection): Deleted.

Canonical link: <a href="https://commits.webkit.org/257358@main">https://commits.webkit.org/257358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/507c9d9fc8ba3d6f33ffce849abcc44d6cbb7ff5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108101 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168353 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85260 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91214 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106029 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6355 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89920 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33362 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88181 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21264 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76288 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1813 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22793 "Found 30 new test failures: http/tests/navigation/javascriptlink-basic.html, http/wpt/cache-storage/cache-quota.any.html, imported/w3c/web-platform-tests/css/css-backgrounds/background-origin-001.html, imported/w3c/web-platform-tests/css/css-flexbox/flexitem-no-margin-collapsing.html, imported/w3c/web-platform-tests/css/css-writing-modes/text-orientation-parsing-001.html, imported/w3c/web-platform-tests/css/cssom-view/cssom-getBoundingClientRect-003.html, imported/w3c/web-platform-tests/fetch/metadata/generated/element-frame.sub.html, imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/javascript-url-global-scope.html, imported/w3c/web-platform-tests/html/cross-origin-opener-policy/popup-same-origin-non-initial-about-blank.https.html, imported/w3c/web-platform-tests/html/cross-origin-opener-policy/reporting/navigation-reporting/reporting-coop-navigated-opener.https.html ... (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1722 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45287 "Found 2 new test failures: fast/forms/select-max-length.html, media/video-inactive-playback.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5051 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6667 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42246 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3114 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/rendering/RenderGrid.h:
```diff
@@ -156,9 +156,16 @@ class RenderGrid final : public RenderBlock {
-    void allocateSpaceForMasonryVectors(Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition, Vector<RenderBox*>& itemsWithIndefinitePosition); 
-    void collectMasonryItems(Grid&, unsigned gridAxisTracksCount, GridTrackSizingDirection masonryDirection, HashMap<RenderBox*, GridArea>& itemsOnFirstTrack, Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition, Vector<RenderBox*>& itemsWithIndefinitePosition) const;
     bool hasDefiniteGridAxisPosition(const RenderBox* child, GridTrackSizingDirection masonryDirection) const;
+    static bool itemGridAreaStartsAtFirstLine(const GridArea& area, GridTrackSizingDirection masonryDirection)
```
Is it possible for these two helpers to be moved as well? They're quite specific to masonry.

Otherwise, appending `InMasonryAxis/Direction` would be good.

The latter I think will be going away once the placement bugs are fixed, so it matters less.

sammygill replied:
Yeah I definitely could! Would these be useful in a non-masonry context in any sort of way? `hasDefiniteGridAxisPosition ` could be changed to `hasDefiniteAxisPosition` so that it could be used to check if an item has a definite span in a specific axis. Otherwise no issues with moving them

mattwoodrow replied:
I'm not sure we have any need for `hasDefiniteAxisPosition` outside of masonry, but either way is fine with me.

sammygill replied:
I'll go ahead and move them unless anyone has any strong opinions otherwise

mattwoodrow reviewed (approved):
This is a really nice cleanup!

Loirooriol reviewed (approved):
Informal LGTM

stwrt commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -166,4 +199,14 @@ LayoutUnit GridMasonryLayout::offsetForChild(const RenderBox& child) const
 
+inline GridTrackSizingDirection GridMasonryLayout::gridAxisDirection() const
+{
+    return m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
```
Would like a comment here to mention that the masonry and grid axis are perpendicular.

sammygill replied:
Done let me know if you want it changed at all

stwrt replied:
Looks good.

stwrt commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -787,15 +787,7 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
-        currentGrid().setupGridForMasonryLayout();
-        populateExplicitGridAndOrderIterator();
-        m_masonryLayout.performMasonryPlacement(firstTrackItems, itemsWithDefiniteGridAxisPosition, itemsWithIndefinitePosition, masonryAxisDirection);
+        m_masonryLayout.performMasonryPlacement(gridAxisTracksBeforeAutoPlacement, masonryAxisDirection);
```
Can we get rid of the masonry lambda here as discussed yesterday?

```CPP
performAutoPlacement();

if(isMasonryRows(style() || isMasonryColumns(style()))
{
  // Do the masonry stuff here
}
  

```

sammygill replied:
I tried to get this done but it looks like I can't completely do it yet because we are reliant on `unsigned gridAxisTracksBeforeAutoPlacement = currentGrid().numTracks(gridAxisDirection);` being computed before auto placement. I tried to see if using `GridPositionsResolver::explicitGridRowCount`/GridPositionsResolver::explicitGridColumnCount, but it doesn't work in all scenarios (e.g. masonry-item-placement-002). After talking with Matt, we shouldn't need this variable anyways after we implement the logic in the implicit grid section (particularly the part about ignoring `grid-auto-flow`), so maybe we can do it then

stwrt replied:
Thanks for taking a look at this.

Not a huge deal for this patch.

sammygill committed:
Move masonry logic from RenderGrid to GridMasonryLayout.
https://bugs.webkit.org/show_bug.cgi?id=248621
rdar://102871735

Reviewed by Tim Nguyen.

This is a somewhat small refactor that involes moving some masonry
specific logic from RenderGrid to GridMasonryLayout. The intial goal was
to remove only collectMasonryItems so that GridMasonryLayout could
hold onto the items directly, but it became pretty clear that some of
the other logic in the surrounding area could also be moved. This should
hopefully have 2 benefits:

1.) Make RenderGrid much more simple and easier to read when we need to
do a masonry layout (since much of the responsibility should be moved)
2.) Allow GridMasonryLayout to much more easily query the masonry item
data structures without having to pass references around everywhere.
This is especially useful if we ever need to check if an item is a
first track item.

There should be no functionality change with this patch.

* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::performMasonryPlacement):
(WebCore::GridMasonryLayout::collectMasonryItems):
(WebCore::GridMasonryLayout::allocateCapacityForMasonryVectors):
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
(WebCore::GridMasonryLayout::nextMasonryPositionForItem):
(WebCore::GridMasonryLayout::gridAxisDirection const):
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::allocateSpaceForMasonryVectors): Deleted.
(WebCore::itemGridAreaStartsAtFirstLine): Deleted.
(WebCore::itemGridAreaIsWithinImplicitGrid): Deleted.
(WebCore::RenderGrid::collectMasonryItems const): Deleted.
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridPositionsResolver.h:
(WebCore::GridPositionsResolver::gridAxisDirection): Deleted.

Canonical link: https://commits.webkit.org/257358@main
